addons/binding/org.openhab.binding.miio/src/main/java/org/openhab/binding/miio/internal/transport/MiIoAsyncCommunication.java

```diff
@@ -0,0 +1,399 @@
+     * the {@link XiaomiSocket} is being closed.
+     *
+     * @param listener - {@link XiaomiSocketListener} to be unregistered
+     */
+    public synchronized void unregisterListener(MiIoMessageListener listener) {
+        getListeners().remove(listener);
+    }
+
+    public int queueCommand(MiIoCommand command) throws MiIoCryptoException, IOException {
+        return queueCommand(command, "[]");
+    }
+
+    public int queueCommand(MiIoCommand command, String params) throws MiIoCryptoException, IOException {
+        return queueCommand(command.getCommand(), params);
+    }
+
+    public int queueCommand(String command, String params)
+            throws MiIoCryptoException, IOException, JsonSyntaxException {
+        try {
+            JsonObject fullCommand = new JsonObject();
+            int cmdId = id.incrementAndGet();
+            fullCommand.addProperty("id", cmdId);
+            fullCommand.addProperty("method", command);
+            fullCommand.add("params", parser.parse(params));
+            MiIoSendCommand sendCmd = new MiIoSendCommand(cmdId, MiIoCommand.getCommand(command),
+                    fullCommand.toString());
+            concurrentLinkedQueue.add(sendCmd);
+            logger.debug("Command added to Queue {} -> {} (Device: {} token: {} Queue: {})", fullCommand.toString(), ip,
+                    Utils.getHex(deviceId), Utils.getHex(token), concurrentLinkedQueue.size());
+            if (needPing) {
+                sendPing(ip);
+            }
+            return cmdId;
+        } catch (JsonSyntaxException e) {
+            logger.warn("Send command '{}' with parameters {} -> {} (Device: {}) gave error {}", command, params, ip,
+                    Utils.getHex(deviceId), e.getMessage());
+            throw e;
+        }
+    }
+
+    MiIoSendCommand sendMiIoSendCommand(MiIoSendCommand miIoSendCommand) {
+        String errorMsg = "Unknown Error while sending command";
+        String decryptedResponse = "";
+        try {
+            decryptedResponse = sendCommand(miIoSendCommand.getCommandString(), token, ip, deviceId);
+            // hack due to avoid invalid json errors from some misbehaving device firmwares
+            decryptedResponse = decryptedResponse.replace(",,", ",");
+            JsonElement response;
+            response = parser.parse(decryptedResponse);
+            if (response.isJsonObject()) {
+                logger.trace("Received  JSON message {}", response.toString());
+                miIoSendCommand.setResponse(response.getAsJsonObject());
+                return miIoSendCommand;
+            } else {
+                errorMsg = "Received message is invalid JSON";
+                logger.debug("{}: {}", errorMsg, decryptedResponse);
+            }
+        } catch (MiIoCryptoException | IOException e) {
+            logger.warn("Send command '{}'  -> {} (Device: {}) gave error {}", miIoSendCommand.getCommandString(), ip,
```

Comment:
At least for IOException, this should be debug only as I am seeing this also for "No route to host" and "Host is down" exceptions, which should simply make the Thing to go OFFLINE, but not do any warn logging.

Reply:
changed it to debug